### ISSUE TITLE
Evil combination of regexes, lexicals, and autovivication

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9806]
Migrated from [rt.perl.org#68020](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=68020) (status was 'resolved')

Searchable as RT68020$

**Comment by p5pRT** [comment]
### From @JohnPeacock

#### Created by jpeacock@cpan.org



Schwern and I simultaneously discovered that some hash tests were 
failing in Damian's Regex&#8203;::Grammars\.  But it turns out to be some 
insideous combination of regexes\, lexicals\, and autovivication and not 
depend on R&#8203;::G at all\.  Damian came up with the following script&#8203;:

&nbsp;   use 5\.010;

&nbsp;   my $rx = qr\{ \(?\{ @&#8203;arr = \[1\]; $arr\[\-1\]\[0\] = 0; my $var; \}\) \}x;

&nbsp;   \#my $dummy1 = '';
&nbsp;   \#my $dummy2 = '';

&nbsp;   my $line = "";
&nbsp;   $line =~ $rx;

&nbsp;   say 'fail' if \!defined $line;

If I run it as is\, I get a segfault\.  If I uncomment one of the
$dummy assignments\, it passes\.  If I uncomment both $dummy assignments\,
I get&#8203;:

&nbsp;   Modification of a read\-only value attempted at test\.pl line 9\.

On the other hand\, Damian got a completely different sort of behavior&#8203;:

> Interestingly\, injecting one \*extra\* lexical declaration anywhere in the
> code \(outside the \(?\{\.\.\.\}\) that is\) has no effect\. But injecting \*two\*
> lexicals anywhere in the code also stops the match from undefinng $line\.
>
> Whereas injecting a second lexical inside the \(?\{\.\.\.\}\) is giving me a
> bus error\.

I told Damian I would file this bug and try and answer any questions
\(he's in full conference travel mode\)\.

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=high

Site configuration information for perl 5.10.0:

Configured by Message Systems, Inc. at Mon Jan 12 16:03:50 EST 2009.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
   Platform:
     osname=linux, osvers=2.6.18-92.1.22.el5, 
archname=x86_64-linux-thread-multi
     uname='linux jpeacock 2.6.18-92.1.22.el5 #1 smp tue dec 16 11:57:43 
est 2008 x86_64 x86_64 x86_64 gnulinux '
     config_args='-des -Uusedtrace -Dmyhostname=localhost 
-Dperladmin=root@localhost -Dcc=/opt/msys/3rdParty/bin/ccache gcc 
-Accflags=-O2 -g -Aldflags= -Alddlflags= -shared -Dcf_by=Message 
Systems, Inc. -Dlibpth=/lib64 /usr/lib64 -Darchname=x86_64-linux 
-Duseshrplib -Di_shadow -Ulocincpth 
-Dlocincpth=/opt/msys/3rdParty/include -Uloclibpth 
-Dloclibpth=/opt/msys/3rdParty/lib -Ui_gdbm -Dusethreads -Duselargefiles 
-Dd_semctl_semun -Di_db -Ui_ndbm -Di_syslog -Dman3ext=3pm 
-Dman1dir=/opt/msys/3rdParty/man/man1 
-Dman3dir=/opt/msys/3rdParty/man/man3 -Duseperlio -Uinstallusrbinperl 
-Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr 
-Dusemultiplicity -Dvendorprefix=/opt/msys -Dprefix=/opt/msys/3rdParty'
     hint=recommended, useposix=true, d_sigaction=define
     useithreads=define, usemultiplicity=define
     useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
     use64bitint=define, use64bitall=define, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='/opt/msys/3rdParty/bin/ccache gcc', ccflags ='-D_REENTRANT 
-D_GNU_SOURCE -O2 -g -fno-strict-aliasing -pipe 
-I/opt/msys/3rdParty/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 
-I/usr/include/gdbm',
     optimize='-O2',
     cppflags='-D_REENTRANT -D_GNU_SOURCE -O2 -g -fno-strict-aliasing 
-pipe -I/opt/msys/3rdParty/include -I/usr/include/gdbm'
     ccversion='', gccversion='4.1.2 20071124 (Red Hat 4.1.2-42)', 
gccosandvers=''
     intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
     ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='/opt/msys/3rdParty/bin/ccache gcc', ldflags =' '
     libpth=/lib64 /usr/lib64
     libs=-lnsl -lgdbm -ldl -lm -lcrypt -lutil -lpthread -lc
     perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
     libc=/lib/libc-2.5.so, so=so, useshrplib=true, libperl=libperl.so
     gnulibc_version='2.5'
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E 
-Wl,-rpath,/opt/msys/3rdParty/lib/perl5/5.10.0/x86_64-linux-thread-multi/CORE'
     cccdlflags='-fPIC', lddlflags='  -shared'

Locally applied patches:



@INC for perl 5.10.0:
     /opt/msys/3rdParty/lib/perl5/site_perl/5.10.0/x86_64-linux-thread-multi
     /opt/msys/3rdParty/lib/perl5/site_perl/5.10.0
     /opt/msys/lib/perl5/vendor_perl/5.10.0/x86_64-linux-thread-multi
     /opt/msys/lib/perl5/vendor_perl/5.10.0
     /opt/msys/lib/perl5/vendor_perl
     /opt/msys/3rdParty/lib/perl5/5.10.0/x86_64-linux-thread-multi
     /opt/msys/3rdParty/lib/perl5/5.10.0
     .


Environment for perl 5.10.0:
     HOME=/home/jpeacock
     LANG=en_US.UTF-8
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
 
PATH=/home/jpeacock/bin:/home/jpeacock/bin:/usr/lib64/qt-3.3/bin:/opt/msys/bin:/usr/kerberos/bin:/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/usr/NX/bin
     PERL_BADLANG (unset)
     SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Jul 29\, 2009 at 11&#8203;:46&#8203;:39AM \-0700\, John Peacock wrote&#8203;:

> Schwern and I simultaneously discovered that some hash tests were 
> failing in Damian's Regex&#8203;::Grammars\.  But it turns out to be some 
> insideous combination of regexes\, lexicals\, and autovivication and not 
> depend on R&#8203;::G at all\.  Damian came up with the following script&#8203;:
> 
>    use 5\.010;
> 
>    my $rx = qr\{ \(?\{ @&#8203;arr = \[1\]; $arr\[\-1\]\[0\] = 0; my $var; \}\) \}x;
> 
>    \#my $dummy1 = '';
>    \#my $dummy2 = '';
> 
>    my $line = "";
>    $line =~ $rx;
> 
>    say 'fail' if \!defined $line;

On Linux\, with blead\, valgrind says&#8203;:

==32406== Invalid read of size 8
==32406==    at 0x5DE7BC&#8203;: Perl\_leave\_scope \(scope\.c&#8203;:839\)
==32406==    by 0x67E829&#8203;: S\_regmatch \(regexec\.c&#8203;:5360\)
==32406==    by 0x6702D4&#8203;: S\_regtry \(regexec\.c&#8203;:2346\)
==32406==    by 0x66F238&#8203;: Perl\_regexec\_flags \(regexec\.c&#8203;:2142\)
==32406==    by 0x53C255&#8203;: Perl\_pp\_match \(pp\_hot\.c&#8203;:1355\)
==32406==    by 0x4F5D5C&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2017\)
==32406==    by 0x4492E3&#8203;: S\_run\_body \(perl\.c&#8203;:2282\)
==32406==    by 0x448806&#8203;: perl\_run \(perl\.c&#8203;:2208\)
==32406==    by 0x41F6DC&#8203;: main \(perlmain\.c&#8203;:117\)
==32406==  Address 0x5c640d0 is 8 bytes after a block of size 32 alloc'd
==32406==    at 0x4C2260E&#8203;: malloc \(vg\_replace\_malloc\.c&#8203;:207\)
==32406==    by 0x4F6657&#8203;: Perl\_safesysmalloc \(util\.c&#8203;:94\)
==32406==    by 0x528615&#8203;: Perl\_av\_extend \(av\.c&#8203;:183\)
==32406==    by 0x52A5D0&#8203;: Perl\_av\_store \(av\.c&#8203;:352\)
==32406==    by 0x4A9C50&#8203;: Perl\_pad\_new \(pad\.c&#8203;:208\)
==32406==    by 0x447BA2&#8203;: S\_parse\_body \(perl\.c&#8203;:2011\)
==32406==    by 0x446CE4&#8203;: perl\_parse \(perl\.c&#8203;:1600\)
==32406==    by 0x41F6C7&#8203;: main \(perlmain\.c&#8203;:115\)
==32406==
==32406== Invalid read of size 4
==32406==    at 0x5DE891&#8203;: Perl\_leave\_scope \(scope\.c&#8203;:849\)
==32406==    by 0x67E829&#8203;: S\_regmatch \(regexec\.c&#8203;:5360\)
==32406==    by 0x6702D4&#8203;: S\_regtry \(regexec\.c&#8203;:2346\)
==32406==    by 0x66F238&#8203;: Perl\_regexec\_flags \(regexec\.c&#8203;:2142\)
==32406==    by 0x53C255&#8203;: Perl\_pp\_match \(pp\_hot\.c&#8203;:1355\)
==32406==    by 0x4F5D5C&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2017\)
==32406==    by 0x4492E3&#8203;: S\_run\_body \(perl\.c&#8203;:2282\)
==32406==    by 0x448806&#8203;: perl\_run \(perl\.c&#8203;:2208\)
==32406==    by 0x41F6DC&#8203;: main \(perlmain\.c&#8203;:117\)
==32406==  Address 0x8 is not stack'd\, malloc'd or \(recently\) free'd

Lines 837 to 849 are&#8203;:

&nbsp;    case SAVEt\_CLEARSV&#8203;:
&nbsp;        ptr = \(void\*\)&PL\_curpad\[SSPOPLONG\];
&nbsp;        sv = \*\(SV\*\*\)ptr;

&nbsp;        DEBUG\_Xv\(PerlIO\_printf\(Perl\_debug\_log\,
&nbsp;         "Pad 0x%"UVxf"\[0x%"UVxf"\] clearsv&#8203;: %ld sv=0x%"UVxf"\<%"IVdf"> %s\\n"\,
&nbsp;        PTR2UV\(PL\_comppad\)\, PTR2UV\(PL\_curpad\)\,
&nbsp;        \(long\)\(\(SV \*\*\)ptr\-PL\_curpad\)\, PTR2UV\(sv\)\, \(IV\)SvREFCNT\(sv\)\,
&nbsp;        \(SvREFCNT\(sv\) \<= 1 && \!SvOBJECT\(sv\)\) ? "clear" : "abandon"
&nbsp;        \)\);

&nbsp;        /\* Can clear pad variable in place? \*/
&nbsp;        if \(SvREFCNT\(sv\) \<= 1 && \!SvOBJECT\(sv\)\) \{


I don't know why/how ptr ends up as garbage\. The block it is 1 pointer beyond
was allocated by this bit of pad\.c

&nbsp;    av\_store\(pad\, 0\, NULL\);


I hope this gives someone some insight\, as this is about as far as I can go\.

Nicholas Clark

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Jul 29\, 2009 at 11&#8203;:46&#8203;:39AM \-0700\, John Peacock wrote&#8203;:
>    my $rx = qr\{ \(?\{ @&#8203;arr = \[1\]; $arr\[\-1\]\[0\] = 0; my $var; \}\) \}x;

Any use of lexical vars within \(?\{\.\.\.\}\) is currently deeply broken due to
the braindead way it's implemented \(eval the code\, then keep the eval'd
CV's body around for future use\. Eveything related to pads and closures is
then wrong\)\.

I've had this on my list of things to fix for years\. Got halfway to
a solution\, then maint pumpkingness took over my life\.\.\.

\-\- 
This email is confidential\, and now that you have read it you are legally
obliged to shoot yourself\. Or shoot a lawyer\, if you prefer\. If you have
received this email in error\, place it in its original wrapping and return
for a full refund\. By opening this email\, you accept that Elvis lives\.

**Comment by p5pRT** [comment]
### From @JohnPeacock

Dave Mitchell wrote&#8203;:
> On Wed\, Jul 29\, 2009 at 11&#8203;:46&#8203;:39AM \-0700\, John Peacock wrote&#8203;:
>>    my $rx = qr\{ \(?\{ @&#8203;arr = \[1\]; $arr\[\-1\]\[0\] = 0; my $var; \}\) \}x;
> 
> Any use of lexical vars within \(?\{\.\.\.\}\) is currently deeply broken due to
> the braindead way it's implemented \(eval the code\, then keep the eval'd
> CV's body around for future use\. Eveything related to pads and closures is
> then wrong\)\.

Damian worked around it with a single s/my/local/ in Regexp&#8203;::Grammars\, 
so it looks like we can probably let this fester a little longer then\.\.\.

John

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jul 29 11&#8203;:46&#8203;:39 2009\, john\.peacock@&#8203;havurah\-software\.org wrote&#8203;:
> Schwern and I simultaneously discovered that some hash tests were
> failing in Damian's Regex&#8203;::Grammars\.  But it turns out to be some
> insideous combination of regexes\, lexicals\, and autovivication and not
> depend on R&#8203;::G at all\.  Damian came up with the following script&#8203;:
> 
>    use 5\.010;
> 
>    my $rx = qr\{ \(?\{ @&#8203;arr = \[1\]; $arr\[\-1\]\[0\] = 0; my $var; \}\) \}x;
> 
>    \#my $dummy1 = '';
>    \#my $dummy2 = '';
> 
>    my $line = "";
>    $line =~ $rx;
> 
>    say 'fail' if \!defined $line;
> 
> If I run it as is\, I get a segfault\.  If I uncomment one of the
> $dummy assignments\, it passes\.  If I uncomment both $dummy
> assignments\,
> I get&#8203;:
> 
>    Modification of a read\-only value attempted at test\.pl line 9\.
> 
> On the other hand\, Damian got a completely different sort of behavior&#8203;:
> 
> > Interestingly\, injecting one \*extra\* lexical declaration anywhere in
> the
> > code \(outside the \(?\{\.\.\.\}\) that is\) has no effect\. But injecting
> \*two\*
> > lexicals anywhere in the code also stops the match from undefinng
> $line\.
> >
> > Whereas injecting a second lexical inside the \(?\{\.\.\.\}\) is giving me
> a
> > bus error\.
> 
> I told Damian I would file this bug and try and answer any questions
> \(he's in full conference travel mode\)\.

Was this fixed by 9e103e269?

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Wed Jul 29 11&#8203;:46&#8203;:39 2009\, john\.peacock@&#8203;havurah\-software\.org wrote&#8203;:
> Schwern and I simultaneously discovered that some hash tests were
> failing in Damian's Regex&#8203;::Grammars\.  But it turns out to be some
> insideous combination of regexes\, lexicals\, and autovivication and not
> depend on R&#8203;::G at all\.  Damian came up with the following script&#8203;:
> 
>    use 5\.010;
> 
>    my $rx = qr\{ \(?\{ @&#8203;arr = \[1\]; $arr\[\-1\]\[0\] = 0; my $var; \}\) \}x;
> 
>    \#my $dummy1 = '';
>    \#my $dummy2 = '';
> 
>    my $line = "";
>    $line =~ $rx;
> 
>    say 'fail' if \!defined $line;
> 
> If I run it as is\, I get a segfault\.  If I uncomment one of the
> $dummy assignments\, it passes\.  If I uncomment both $dummy
> assignments\,
> I get&#8203;:
> 
>    Modification of a read\-only value attempted at test\.pl line 9\.
> 
> On the other hand\, Damian got a completely different sort of behavior&#8203;:
> 
> > Interestingly\, injecting one \*extra\* lexical declaration anywhere in
> the
> > code \(outside the \(?\{\.\.\.\}\) that is\) has no effect\. But injecting
> \*two\*
> > lexicals anywhere in the code also stops the match from undefinng
> $line\.
> >
> > Whereas injecting a second lexical inside the \(?\{\.\.\.\}\) is giving me
> a
> > bus error\.
> 
> I told Damian I would file this bug and try and answer any questions
> \(he's in full conference travel mode\)\.

Was this fixed by 9e103e269?

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Jul 16\, 2011 at 06&#8203;:27&#8203;:44AM \-0700\, Father Chrysostomos via RT wrote&#8203;:
> On Wed Jul 29 11&#8203;:46&#8203;:39 2009\, john\.peacock@&#8203;havurah\-software\.org wrote&#8203;:
> > Schwern and I simultaneously discovered that some hash tests were
> > failing in Damian's Regex&#8203;::Grammars\.  But it turns out to be some
> > insideous combination of regexes\, lexicals\, and autovivication and not
> > depend on R&#8203;::G at all\.  Damian came up with the following script&#8203;:
> > 
> >    use 5\.010;
> > 
> >    my $rx = qr\{ \(?\{ @&#8203;arr = \[1\]; $arr\[\-1\]\[0\] = 0; my $var; \}\) \}x;
> > 
> >    \#my $dummy1 = '';
> >    \#my $dummy2 = '';
> > 
> >    my $line = "";
> >    $line =~ $rx;
> > 
> >    say 'fail' if \!defined $line;
> > 
> > If I run it as is\, I get a segfault\.  If I uncomment one of the
> > $dummy assignments\, it passes\.  If I uncomment both $dummy
> > assignments\,
> > I get&#8203;:
> > 
> >    Modification of a read\-only value attempted at test\.pl line 9\.
> > 
> > On the other hand\, Damian got a completely different sort of behavior&#8203;:
> > 
> > > Interestingly\, injecting one \*extra\* lexical declaration anywhere in
> > the
> > > code \(outside the \(?\{\.\.\.\}\) that is\) has no effect\. But injecting
> > \*two\*
> > > lexicals anywhere in the code also stops the match from undefinng
> > $line\.
> > >
> > > Whereas injecting a second lexical inside the \(?\{\.\.\.\}\) is giving me
> > a
> > > bus error\.
> > 
> > I told Damian I would file this bug and try and answer any questions
> > \(he's in full conference travel mode\)\.
> 
> Was this fixed by 9e103e269?

Most likely\.

\-\- 
The warp engines start playing up a bit\, but seem to sort themselves out
after a while without any intervention from boy genius Wesley Crusher\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#17

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
